package/src/lock.rs

```diff
@@ -0,0 +1,267 @@
+//! Lock files and lock file utilities
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    num::ParseIntError,
+    path::{Path, PathBuf},
+};
+
+use gix::ObjectId;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::{Error, IoResultExt},
+    snapshot::Snapshot,
+    Dependency, GitDependency, ManifestFile, Precise,
+};
+
+/// We need to give names to entries in the lock-file, so that we can refer to dependencies.
+///
+/// It's tempting to try to use `Precise` for identifying dependencies, but we can't because
+/// the lock file doesn't store paths on the local filesystem. So we identify packages by
+/// the name that its importer uses, and add numbers to ensure uniqueness.
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    serde_with::SerializeDisplay,
+    serde_with::DeserializeFromStr,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+)]
+pub struct EntryName {
+    pub name: String,
+    pub id: u32,
+}
+
+impl std::fmt::Display for EntryName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.id == 0 {
+            write!(f, "{}", self.name)
+        } else {
+            write!(f, "{} {}", self.name, self.id)
```

Comment:
Naive question: is it possible that a package name has a space in the name?

Reply:
You can't use such a package because the import syntax for packages requires the package name to be an identifier ("identifier" in the sense of the nickel lexer). But it's a good point, `nickel_lang_package` should check this.

Reply:
I've added this check to `std.package.Manifest`